File: bkasm/src/inbuf.c

#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <assert.h>
#include <string.h>
#ifndef WIN32
#ifndef _Windows
#ifndef __MSDOS__
#include <unistd.h>
#endif
#endif
#endif
#include "bkasm.h"
#include "inbuf.h"

static InbufFrame frames[MAX_INCLUDE_DEPTH];
static int frame_count;

static const char* inbuf;
static int bufsize;
static int cursym;
static InbufCurrentString currStr;

InbufCurrentString *inbuf_currstr()
{
    return (InbufCurrentString *) &currStr;
}

static void inbuf_markcurrstr()
{
    int i;
    for (i = 0; i + cursym < bufsize && i < INBUF_STRING_SIZE; i++)
    {
        currStr.str[i] = inbuf[cursym + i];
        if (currStr.str[i] == 10)
            currStr.str[i] = 0;
    }
}

static void inbuf_save_frame(int idx)
{
    frames[idx].buf = inbuf;
    frames[idx].bufsize = bufsize;
    frames[idx].cursym = cursym;
    frames[idx].currStr = currStr;
}

static void inbuf_restore_frame(int idx)
{
    inbuf = frames[idx].buf;
    bufsize = frames[idx].bufsize;
    cursym = frames[idx].cursym;
    currStr = frames[idx].currStr;
}

static void inbuf_pop(void)
{
    int idx = frame_count - 1;
    --frame_count;
    free((void*)frames[idx].buf);
    if (frame_count > 0)
        inbuf_restore_frame(frame_count - 1);
}

unsigned char inbuf_next_char()
{
    unsigned char chr;
    if (cursym < bufsize)
        chr = inbuf[cursym++];
    else if (frame_count > 1)
    {
        inbuf_pop();
        return inbuf_next_char();
    }
    else
    {
        chr = 0xff;
    }
    if (chr == 10)
    {
        currStr.num++;
        inbuf_markcurrstr();
    }
    return chr;
}

void inbuf_init(const char* _inbuf, int _bufsize)
{
    frame_count = 1;
    inbuf = _inbuf;
    bufsize = _bufsize;
    cursym = 0;
    currStr.num = 1;
    inbuf_markcurrstr();
    inbuf_save_frame(0);
}

int inbuf_push_file(const char *filename)
{
    FILE *f;
    long sz;
    char *buf;

    if (frame_count >= MAX_INCLUDE_DEPTH)
        return -1;

    inbuf_save_frame(frame_count - 1);

    f = fopen(filename, "r");
    if (!f)
        return -1;

    fseek(f, 0, SEEK_END);
    sz = ftell(f);
    if (sz < 0)
    {
        fclose(f);
        return -1;
    }
    rewind(f);

    buf = malloc(sz + 1);
    if (!buf)
    {
        fclose(f);
        return -1;
    }

    if (fread(buf, 1, sz, f) != (size_t)sz)
    {
        free(buf);
        fclose(f);
        return -1;
    }
    buf[sz] = '\0';
    fclose(f);

    inbuf = buf;
    bufsize = sz;
    cursym = 0;
    currStr.num = 1;
    inbuf_markcurrstr();

    inbuf_save_frame(frame_count);
    ++frame_count;
    return 0;
}

void inbuf_cleanup(void)
{
    while (frame_count > 1)
    {
        int idx = frame_count - 1;
        --frame_count;
        free((void*)frames[idx].buf);
    }
    frame_count = 0;
}

File: bkasm/csharp/Inbuf.cs
using System;
using System.IO;

namespace Bkasm
{
    public class InbufCurrentString
    {
        public int num;
        public char[] str = new char[Constants.INBUF_STRING_SIZE];

        public InbufCurrentString()
        {
            num = 1;
        }
    }

    public class InbufFrame
    {
        public byte[] buf;
        public int bufsize;
        public int cursym;
        public InbufCurrentString currStr;
    }

    public static class Inbuf
    {
        public const int MAX_INCLUDE_DEPTH = 16;

        private static InbufFrame[] frames = new InbufFrame[MAX_INCLUDE_DEPTH];
        private static int frameCount;
        private static byte[] inbuf;
        private static int bufsize;
        private static int cursym;
        private static InbufCurrentString currStr = new InbufCurrentString();

        public static InbufCurrentString CurrStr()
        {
            return currStr;
        }

        private static void MarkCurrStr()
        {
            for (int i = 0; i + cursym < bufsize && i < Constants.INBUF_STRING_SIZE; i++)
            {
                currStr.str[i] = (char)inbuf[cursym + i];
                if (currStr.str[i] == '\n')
                    currStr.str[i] = '\0';
            }
        }

        private static void SaveFrame(int idx)
        {
            frames[idx] = new InbufFrame();
            frames[idx].buf = inbuf;
            frames[idx].bufsize = bufsize;
            frames[idx].cursym = cursym;
            frames[idx].currStr = new InbufCurrentString();
            frames[idx].currStr.num = currStr.num;
            Array.Copy(currStr.str, frames[idx].currStr.str, Constants.INBUF_STRING_SIZE);
        }

        private static void RestoreFrame(int idx)
        {
            inbuf = frames[idx].buf;
            bufsize = frames[idx].bufsize;
            cursym = frames[idx].cursym;
            currStr.num = frames[idx].currStr.num;
            Array.Copy(frames[idx].currStr.str, currStr.str, Constants.INBUF_STRING_SIZE);
        }

        private static void Pop()
        {
            int idx = frameCount - 1;
            --frameCount;
            frames[idx] = null;
            if (frameCount > 0)
                RestoreFrame(frameCount - 1);
        }

        public static byte NextChar()
        {
            byte chr;
            if (cursym < bufsize)
                chr = inbuf[cursym++];
            else if (frameCount > 1)
            {
                Pop();
                return NextChar();
            }
            else
                chr = Constants.EOF_SYM;

            if (chr == '\n')
            {
                currStr.num++;
                MarkCurrStr();
            }
            return chr;
        }

        public static void Init(byte[] buf, int size)
        {
            frameCount = 1;
            inbuf = buf;
            bufsize = size;
            cursym = 0;
            currStr.num = 1;
            MarkCurrStr();
            SaveFrame(0);
        }

        public static int PushFile(string filename)
        {
            if (frameCount >= MAX_INCLUDE_DEPTH)
                return -1;

            SaveFrame(frameCount - 1);

            byte[] buf;
            try
            {
                buf = File.ReadAllBytes(filename);
            }
            catch
            {
                return -1;
            }

            inbuf = buf;
            bufsize = buf.Length;
            cursym = 0;
            currStr.num = 1;
            MarkCurrStr();

            SaveFrame(frameCount);
            ++frameCount;
            return 0;
        }

        public static void Cleanup()
        {
            while (frameCount > 1)
            {
                int idx = frameCount - 1;
                --frameCount;
                frames[idx] = null;
            }
            frameCount = 0;
        }
    }
}
